### ISSUE TITLE
HTTPS perf issue on Linux/Ubuntu

### ISSUE THREAD

**Issue by seguler** [dotnet/corefx#18173]
We have a sample app that uploads large files into Azure Storage on Linux (Ubuntu 16.10, .NET Core 1.1). When we use HTTP endpoint we get about 600MB/s throughput, whereas as soon as we change to HTTPS, the throughput drops to about 100MB/s. At first we thought this could be a bottleneck on Azure Storage, but after testing the same scenario with curl on Linux, as well as running the test on Windows, we concluded that this issue happens only with .NET Core on Linux when HTTPS is used. 

Using .NET Core on Windows, we get similar throughput around 600MB/s for both HTTP and HTTPS. And on Linux, we also tested this using a curl command and both HTTP and HTTPS provide similar results, hence raising the issue here.

The sample app is here: https://github.com/seguler/azure-storage-throughput-test 

Any pointers on the issue ? 

@JeffreyRichter @polarapfel

**Comment by stephentoub** [comment]
A few questions to try to narrow this down:
- I took a quick look at the sample but don't see HTTP-related types being used directly.  I assume the usage is coming from the Azure SDK?
- Is it possible to come up with a repro that uses HttpClient directly?
- What version of .NET Core are you running against (i.e. which version of System.Net.Http.dll)?  Have you tried using .NET Core 2.0 (not yet released)?  There have been some measurable perf improvements made both between 1.0->1.1 and 1.1->2.0, in particular on Linux.
- Do you know if any custom work is being done with certificates?  I ask because if you don't configure, for example, a remote cert validation callback, don't turn on cert revocation checking, etc., basically just leave certificate-related functionality with the defaults, then I would expect any perf difference between http and https with HttpClient to match a difference between http and https with using curl directly; once you start configuring cert-related functionality, we have to start doing our own logic with X509Certificates, so there could be an issue there, e.g. https://github.com/dotnet/corefx/commit/5036cc84488a87cf2c42b4895a4e0ebad9d58a62

cc: @bartonjs, @geoffkizer

**Comment by seguler** [comment]
- We are using the Azure Storage SDK with .NET Core 1.1. It seems System.IO version is 4.3.0. I have tested 1.0 as well and saw no difference.
- I couldn't possibly recreate this in a simple unit test with HttpClient directly, as we have a lot of parallelism - as well as different REST API calls- in order to get to 600MB/s in the Storage SDK. However the HttpClient in the SDK is here, if it will help: https://github.com/Azure/azure-storage-net/tree/master/Lib/ClassLibraryCommon/Shared/Protocol
- How can I upgrade to .Net Core 2.0 ? Is it published in one of the Ubuntu package repos ?

**Comment by seguler** [comment]
In response to the custom work with certs question, we don't do any custom work with certificates. It is pretty straight forward: https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Shared/Protocol/HttpWebRequestFactory.cs

**Comment by stephentoub** [comment]
> How can I upgrade to .Net Core 2.0 ? Is it published in one of the Ubuntu package repos ?

https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md
@danmosemsft, @karelz, are these the latest and greatest docs available?

> I have tested 1.0 as well and saw no difference.

Not much happened here between 1.0 and 1.1.  We did do some work on performance, at least for some scenarios, for 2.0, in particular for Unix, so it'd be good to try out the latest bits to see whether it touches on your case.

> It is pretty straight forward

Ooh, you're using HttpWebRequest?  If you care about performance, first thing to do is to stop using HttpWebRequest and switch to using HttpClient.  In desktop, HttpClient is implemented on top of HttpWebRequest.  In core, it's the inverse, with HttpWebRequest implemented on top of HttpClient, and due to a mismatch in how functionality is exposed between the two surface areas, HttpWebRequest in core is fairly inefficient.  Specifically, HttpClient is optimized for use by many requests, and so things like connection pooling are tied to an individual HttpClient instance.  In core currently, every HttpWebRequest ends up creating its own HttpClient, which means not only are you paying the overheads of spinning up a new HttpClient instance for every HttpWebRequest, you're losing out on things like connection pooling.  It would be better to switch your code to using a single HttpClient instance (the same instance can be used by any number of requests concurrently, so you can create, cache, and reuse a single HttpClient instance).  I don't know if that'll help your particular case, but in general I'd expect it to be much better.

> on Linux (Ubuntu 16.10

Which version of libcurl do you have? e.g. what's the output from `curl --version`?  Things work when the version of libcurl being used is built against OpenSSL as its backend.  I know in the past I saw some significant perf degradation when it was instead built against other libraries like GnuTLS.  If you see something like NSS and or GnuTLS listed in `curl --version`, it might be worth trying to install the OpenSSL-based libcurl and see if that has any effect on throughput, at least as an experiment.

**Comment by JohnRusk** [comment]
@stephentoub I'm also experiencing the same problem.  I can't quite see how HttpWebRequest/vis HttpClient can be the answer.  Because:

1. @seguler  and I both observe the performance difference when switching between HTTP and HTTPS - _both_ of those are using HttpWebRequest.  So the difference appears to be not whether one is using HttpWebRequest or HttpClient, but whether one is using HTTP or HTTPS.
2. Do the differences you describe really account for such a large difference - e.g. a factor of 6 in @seguler's tests?  My own gut feel is that there's something more going on here, more than just what you mention in your reply of April 21.
3. I'm using a library that I don't control.  That library is calling HttpWebRequest.  So I can't change to use HttpClient instead.  

Besides, isn't the point of .NET Core to allow some degree of cross platform compatibility? Therefore, if a cross platform library works fine on Windows, but performs unusably slowly on Linux, isn't that a bug in .NET Core?

**Comment by stephentoub** [comment]
> So the difference appears to be not whether one is using HttpWebRequest or HttpClient, but whether one is using HTTP or HTTPS.

I understand the difference you're seeing is between http and https. My point was that some of the costs associated with https are at the opening of the connection, so you're going to see higher overheads to https vs http if you make a single request per connection than if you reuse a connection for multiple requests. And since pooling is done per HttpClientHandler instance and this based on the current implementation per HttpWebRequest instance, there's no pooling with HWR, and thus there's no amortization of those costs. I'm in no way saying that's a good thing, simply explaining the state of the world.

> Do the differences you describe really account for such a large difference

Potentially. I've not been able to repro so I've not been able to investigate.

> That library is calling HttpWebRequest. So I can't change to use HttpClient instead.

Understood.

> isn't that a bug in .NET Core?

It's quite possible there's a bug, . If so, it's possible it's in .NET Core itself, or in libcurl, or in OpenSSL, or somewhere else. Never said otherwise. If you can help with a simple, isolated repro we can run to better understand and investigate, we'd be happy to do so.

**Comment by JohnRusk** [comment]
Thanks.
That makes sense, re amortization of connection set up.  I'm using the same library as @seguler and, like him, I'm not sure I can easily make a repro without that library.  However, I think my first step should probably to be to test on the current pre-release version of .NET Core 2.0 and see if that makes any difference. If it doesn't, I'll probably have to see if I can find some may to make a usable repro....

**Comment by stephentoub** [comment]
Thanks, that's helpful. We're also looking more into SSL performance for 2.1, so we'll be doing some investigations as part of that. I do have to say I'm surprised to hear that the https throughout on any platform was equivalent to http... generally even for steady-state throughout SSL adds some measurable overhead. Maybe there are some other factors in the workload overshadowing that.

**Comment by JohnRusk** [comment]
@stephentoub I have tested on 2.0.0. preview 2, and the problem still exists.   cc @seguler 

Re difference between HTTP and HTTPS: in other related work, I'm testing TCP connections (no HTTP) both with and without TLS encryption.  There's a clear difference in CPU usage, but not much difference in throughput.

**Comment by EmmaZhu** [comment]
@stephentoub I just tested perf for HTTP/HTTPS on .Net Core 2.0.0. Didn't see large difference on perf between HTTP and HTTPS. Found a phenomenon that would be worth to take a look.

The test I've done is to send data to Azure Storage with HttpClient. Code I used to send data is following:
```
                    using (HttpContent httpContent = new StreamContent(new MemoryStream(buffer)))
                    {
                        httpContent.Headers.Add("x-ms-client-request-id", Guid.NewGuid().ToString());
                        httpContent.Headers.Add("x-ms-version", "2016-05-31");
                        using (HttpResponseMessage response = await httpClient.PutAsync(
                            string.Format("{0}://emmahk2.blob.core.windows.net/container1/***",
                            protocol,
                            blockId),
                            httpContent))
                        {
                            response.EnsureSuccessStatusCode();
                        }
                    }
```

I tried to upload 128 * 4MB to a blob in Azure Storage. It's much faster to call the above code sequentially than in parallel (64 or 128 async takes started at the same time). From fiddler, I found that 

- when the code was called sequentially, the request was sent one by one immediately. It took about 20s to complete all requests.

- When the code is invoke in parallel, it took very long time before sending the first request, then all request/response were handled in seconds. At last, it took 100+ seconds at total to complete all requests.

To fully use network bandwidth to upload data as fast as possible to Azure Storage, we do uploading in large parallel, while seems it doesn't work as what we expected. Could you help to take a look? What would you suggest us to do to improve performance?

**Comment by EmmaZhu** [comment]
@stephentoub I've just done some test on .Net Core 2.0.0 on sending data to Azure Storage with HttpClient. 

For normally the ingress to Azure Storage on single blob won't be greater than 100MB, the speed with HTTP and HTTPS are very similar when it's lower than 100MB. When I try to upload to 16 blobs at the same time, ingress for HTTP could be around 400MB/s, while ingress for HTTPS is only 130+MB/s.

 I also tried the same with .Net Core 2.0 on Windows, the ingress could reach 800MB/s at most for both HTTP and HTTPS. This speed could be caused by Azure Storage's adjustment, for the test on Windows is after the one for Linux. 

I can share you my code if you'd want a repro.

**Comment by JohnRusk** [comment]
@stephentoub What's the status of this?  Have you tried running Emma's code to repro the issue?  Is there anything else that Emma, Sercan or I could do to help progress this issue?   I ask because it is potentially significant to everyone who needs to upload significant quantities of data from Linux to Azure.

**Comment by stephentoub** [comment]
Sorry for the delay here; this fell off my radar.  I will aim to try out the repro tomorrow.

**Comment by JohnRusk** [comment]
Thanks @stephentoub .  We'll be interested to hear whether it's reproducibly for you.

**Comment by stephentoub** [comment]
I'd chatted about this offline last week with Emma and folks.  Here's the relevant portion of the info I shared...

---

"Here’s what I see (I added min/max/average to Emma’s repro, all values are in milliseconds):
```
~/stoubtest$ dotnet run -c Release http false
14716
Min: 183
Max: 2075
Average: 893.99609375

~/stoubtest$ dotnet run -c Release https false
60686
Min: 637
Max: 5696
Average: 3704.4140625
```
After some investigation, I believe what’s happening is that we’re effectively being limited by the architecture of libcurl.  Essentially, libcurl operates on a single-threaded model.  It provides a model whereby you create a “multi” handle, which is the thing that manages all of the resources, the connection pool, etc., and you can queue multiple “easy” handles (requests) to run on that multi handle concurrently.  However, the multi handle itself only allows code to access it from a single thread at a time, which means that although you can have many outstanding requests/response on a single multi handle at a time and they can all be concurrently going across the network, the actual compute processing of work associated with any of those requests is serialized.  When we’re talking about https, that also means that all of the work associated with encrypting/decrypting the request/response is serialized, as that all happens as part of processing work associated with the multi handle.  The repro VM Emma gave me access to has 16 logical cores, but `top` shows that we’re only maxing out one of them.

You could try working around it by going slightly against the typical `HttpClient` advice, and instead of having just one `HttpClient` instance for all requests, create N `HttpClient` instances and load balance your requests across them, e.g. where N == `Environment.ProcessorCount`.  Each `HttpClient` instance (or more specifically each `HttpClientHandler` instance) on Unix is tied to a single `CurlHandler` instance which is tied to a single multi handle, so by creating N instead of 1, you’ll end up with N multi handles that can all run independently and all be processed independently.  They each have their own connection pool, and they each can be processed by a single thread at a time, which means you can have N threads all doing work simultaneously (for different requests).  I just tried making a small modification to the repro program Emma sent, to simply round-robin through `Environment.ProcessorCount` `HttpClients` rather than always using the same 1, and here’s what I got:
```
~/stoubtest$ dotnet run -c Release http false
9820
Min: 136
Max: 2514
Average: 582.6513671875

~/stoubtest$ dotnet run -c Release https false
11279
Min: 146
Max: 3681
Average: 665.169921875
```
At one point we experimented with doing this under the covers in `CurlHandler` itself, basically having `CurlHandler` itself maintain `~Environment.ProcessorCount` multi handles, and employ various strategies for partitioning the work between them.  We found that for some workloads it worked out really well, but for others it made things worse.  It also makes implementing some of the throttling controls exposed via HttpClientHandler challenging. We could certainly revisit it, though I’m not entirely confident we’d reach a better answer than last time; still worth taking another look at, though.

For .NET Core 2.1, we’ve been working on a new managed replacement for `CurlHandler`.  I just tried Emma’s repro with just a single `HttpClient` again, and here’s what I get using ManagedHandler (using 2.1 bits from corefx master, plus a one-line hack to work around a known scalability bug in the master src):
```
~/corefx21bits$ export COMPlus_UseManagedHttpClientHandler=true 
~/corefx21bits$ ./corerun stoubtest.dll https false
10833
Min: 159
Max: 2192
Average: 621.81640625
```
so once this is available, it should also address the perf issue, and we’re still actively working on the performance here and expect it to get better still."

**Comment by davidsh** [comment]
Please try out .NET Core 2.1 Preview 2.  It has the new SocketsHttpHandler as the default HTTP stack.  It is much better in performance!

https://blogs.msdn.microsoft.com/dotnet/2018/04/11/announcing-net-core-2-1-preview-2/

If you are having issues using the new .NET Core 2.1, please open a new issue. Thanks!
